sqldelight-idea-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/actions/file/SqlDelightCreateFileAction.kt

```diff
@@ -0,0 +1,186 @@
+import java.util.Properties
+
+/**
+ * Creates a new SqlDelight file/table/migration from a template (see [fileTemplates.internal])
+ *
+ * Adopted from [NewKotlinFileAction](https://github.com/JetBrains/kotlin/blob/f6a739bbc5b66a8f2cb5cb61acb1e87c899fa11e/idea/src/org/jetbrains/kotlin/idea/actions/NewKotlinFileAction.kt#L56)
+ */
+class SqlDelightCreateFileAction : CreateFileFromTemplateAction(
+    CAPTION,
+    "Creates new SqlDelight file or table",
+    SqlDelightFileType.icon
+), DumbAware {
+
+  override fun postProcess(
+    createdElement: PsiFile?,
+    templateName: String?,
+    customProperties: MutableMap<String, String>?
+  ) {
+    super.postProcess(createdElement, templateName, customProperties)
+
+    if (createdElement is SqlDelightFile) {
+      val isTableTemplate = PsiTreeUtil.collectElements(createdElement) { it is SqlTableName }.isNotEmpty()
+      val editor = FileEditorManager.getInstance(createdElement.project).selectedTextEditor ?: return
+      if (editor.document == createdElement.viewProvider.document) {
+        if (isTableTemplate) {
+          // move caret inside "create table" sql statement
+          val lineCount = editor.document.lineCount
+          if (lineCount > 0) {
+            editor.caretModel.moveToLogicalPosition(LogicalPosition(lineCount - 2, 0))
+          }
+        } else {
+          // move caret after "create table", so the user can enter the table name conveniently
+          val tableKeyword = PsiTreeUtil.collectElements(createdElement) { it.node.elementType == TABLE }.first()
+          editor.caretModel.moveToOffset(tableKeyword.textRange.endOffset + 1)
+        }
+      }
+    }
+  }
+
+  override fun getActionName(directory: PsiDirectory?, newName: String, templateName: String?): String = CAPTION
+
+  override fun buildDialog(project: Project?, directory: PsiDirectory?, builder: CreateFileFromTemplateDialog.Builder) {
+    val icon = iconProvider.getIcon(dummyFile, 0, project)
+    builder.setTitle("New $CAPTION")
+        .addKind("File", icon, "SqlDelight File")
+        .addKind("Table", icon, "SqlDelight Table")
+        .addKind("Migration", icon, "SqlDelight Migration")
+        .setValidator(NameValidator)
+  }
+
+  override fun isAvailable(dataContext: DataContext): Boolean {
+    if (!super.isAvailable(dataContext)) return false
+
+    val project = CommonDataKeys.PROJECT.getData(dataContext)!!
+    val file = CommonDataKeys.VIRTUAL_FILE.getData(dataContext)!!
+    val propertiesFolder = project.propertiesFolder ?: return false
+
+    val props = VfsUtil.collectChildrenRecursively(propertiesFolder)
+        .find { it.name == SqlDelightPropertiesFile.NAME } ?: return false
+    val propsFile = SqlDelightPropertiesFile.fromText(props.inputStream.reader().readText())!!
+
+    // only show the SqlDelight.New option if the selected folder is under src/**/sqldelight or a custom one
+    return propsFile.databases.first().compilationUnits
```

Comment:
This means it's available only when you right-click on one of the `sourceFolders` or its subdirectories (`sqldelight` by default) - not sure if it's the best way, or should we show the action somewhere up in the hierarchy (e.g. `src`)?

Reply:
this is really clever, and i think exactly the functionality we want. nice